packages/next/src/server/app-render.tsx

```diff
@@ -1138,7 +1138,10 @@ export async function renderToHTMLOrFlight(
         ...(isPage && searchParamsProps),
       }
 
-      await collectMetadata(tree, layerProps, metadataItems)
+      // If the layout is located next to the page as leaf layer
+      const isLeafLayout =
+        typeof parallelRoutes.children?.[2].page !== 'undefined'
```

Comment:
I tried this out and I believe it doesn't quite work either

```
// scenario 1 
/app/layout.js
/app/page.js
/app/deep/page.js

visit "/" , Layout should be leaf and is leaf
visit "/deep", Layout should NOT be leaf, but it is based on this implementation
```

Reply:
I think this can be accomplished by checking if the path's up to the filename are the same between the potentially leaveLayout and the page

something like (in psuedocode)
```
parallelRoutes.children?.[2].page?.[1].sliceEverythingExceptTheFilename() === currentLayoutsPathWithoutFilename()
```